### PR TITLE
Proposal for change of CPython supported versions

### PR DESCRIPTION
**Pull request by albanD**
The main change here is to extend support until each cpython version is EOL and also clarify timeline for this yearly process.

Rendered version: https://github.com/pytorch/rfcs/blob/rfc-38/RFC-0038-cpython-support.md

### CODE REVIEW

**Comment by isuruf** [comment]
Is this proposal just for pytorch or other repos as well? I ask because `torchfix` has already dropped support for python 3.8 while pytorch has not, leading to issues with developing on python 3.8 environment. (We use a python 3.8 dev environment to make it easier to not introduce any features not supported by 3.8)

**Comment by albanD** [comment]
This is definitely focused on pytorch/pytorch and other repos can have a limit be higher if that makes sense for their userbase.
For torchfix, I would expect it supports the same versions as PyTorch, cc @kit1980 maybe you have more details on this?

File: RFC-0038-cpython-support.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+
+# [CPython version support]
+
+**Authors:**
+* @albanD
+
+## **Motivation**
+
+CPython release schedule is regularly cadenced every year. Given that the amount of work to enable and sunset versions yearly is fixed, we should be proactive in how we handle these tasks.
```
Comment (malfet) [inline RFC-0038-cpython-support.md:9]: Nit (imo easier to parse than "regularly cadenced")
```suggestion
CPython follows annual release cycle. Given that the amount of work to enable and sunset versions yearly is fixed, we should be proactive in how we handle these tasks.
```

File: RFC-0038-cpython-support.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  - When: Once the last supported release happened
+  - ETA: 2 weeks
+  - Goal: remove all wheel and conda binaries for the EOL version
+  - Who: Infra
+- Upgrade oldest version CI
+  - When: After nightlies are not published anymore
+  - ETA: 2 weeks
+  - Goal: migrate all the CI jobs running on the oldest version to the soon-to-be oldest version
+  - Who: Infra
+
```
Comment (malfet) [inline RFC-0038-cpython-support.md:102]: Would be nice to add a paragraph here about BE work associated with removing workarounds specific to older version (i.e. if any `if sys.version < (x, y)` in Python code and similar ifdefs in C++)

**Review by malfet — APPROVED** [review]
LGTM, but please consider adding a stretch goal about BE work associated with deprecating older versions (associated with some typing restrictions, use of walrus operator, etc)

**Review by atalman — APPROVED** [review]

**Comment by kit1980** [comment]
For TorchFix we depend on LibCST , which requires at least Python 3.9 now.
But it's kinda irrelevant because while is 3.9 required run TorchFix, the code to verify can be any of Python 3.0 - 3.12.

**Pull request merged**